### PR TITLE
[FEAT] Auth, Login, create account with username and password

### PR DESCRIPTION
**Pull request by dni**
# new login and accounts based authentication

- introducing session based auth for browsers and and jwt bearer access-tokens for api
- backwards compatible with, ?usr get params and x-api-key

## notes
- there should be a road to drop support for the old authentications style or some kind of new version without that.
- user.id is encoded inside the bearer token

## todo

- [x] remove usr from nodemanagment
- [x] fix 3 failing tests

### CODE REVIEW

**Comment by bitkarrot** [comment]
looking like good stuff here @dni 

can we haz Nostr login? or maybe later.....  :D

**Comment by motorina0** [comment]
This PR uses [fastapi_login](https://github.com/tiangolo/fastapi/tree/master/fastapi/security) which is a 3rd party library.
At the same time [fastap.security](https://github.com/MushroomMaula/fastapi_login) offers the same functionality.

I believe we need a good reason for bringing in a new library instead of using the standard one.

**Comment by motorina0** [comment]
There should be a field in settings: `lnbits_allowed_auth_methods=`. Possible values:
 - `user_id` (default, what it is now. Necessary for backwards compability)
 - `email` (email and password)
 - `nostr` (NIP-07, browser extension recommended)
 - `google` (supported by OAuth2)
 - `twitter` (supported by OAuth2)
 - ...

Eg:
```env
# lnbits_allowed_auth_methods=user_id
# lnbits_allowed_auth_methods=email
lnbits_allowed_auth_methods=email,nostr,twitter

```

**Comment by jackstar12** [comment]
> This PR uses [fastapi_login](https://github.com/tiangolo/fastapi/tree/master/fastapi/security) which is a 3rd party library. At the same time [fastap.security](https://github.com/MushroomMaula/fastapi_login) offers the same functionality.
> 
> I believe we need a good reason for bringing in a new library instead of using the standard one.

fast api login uses fastapi security under the hood. Its just more convenient.
Just like using sqlalchemy can be considered more convenient than using psycopg2 directly.

**Comment by dni** [comment]
> There should be a field in settings: `lnbits_allowed_auth_methods=`. Possible values:
> 
>     * `user_id` (default, what it is now. Necessary for backwards compability)
> 
>     * `email` (email and password)
> 
>     * `nostr` (NIP-07, browser extension recommended)
> 
>     * `google` (supported by OAuth2)
> 
>     * `twitter` (supported by OAuth2)
> 
>     * ...
> 
> 
> Eg:
> 
> ```dotenv
> # lnbits_allowed_auth_methods=user_id
> # lnbits_allowed_auth_methods=email
> lnbits_allowed_auth_methods=email,nostr,twitter
> ```

its a good idea, but it exceeds the scope of this pr. 
this pr is meant to get rid of the ?usr inside url.  we can build upon this PR further down the line.

**Comment by motorina0** [comment]
> fast api login uses fastapi security under the hood. Its just more convenient - but I dont know about the exact differences, Just like using sqlalchemy can be considered more convenient than using psycopg2 directly.

Ok, but any 3rd party library has some risk&disadvantages compared with the base framework, some examples:
 - it will not be maintained after a while
 - fewer people know how to use it
 - it does not allow access to some low-level APIs
 - footprint (an extra installed library)
 - maintenance (extra lib to upgrade)
 - ...

I'm not saying A or B is better, but if A is the built-in default and we go with B then we need some good reasons to do so.

**Comment by motorina0** [comment]
> its a good idea, but it exceeds the scope of this pr.
> this pr is meant to get rid of the ?usr inside url. we can build upon this PR further down the line.

- it is important to maintain backwards compatibility, some people will not want to switch to `email+pass` login right away
- other users might want to ditch immediately `user_id` only auth and only use `email+pass`, they need to specify this somehow
- I agree that `nostr`, `google`, `twitter` can come with a second PR

File: lnbits/core/views/api.py [hunk 1]
```diff
@@ -135,12 +135,9 @@ async def api_wallet(wallet: WalletTypeInfo = Depends(get_key_type)):
...
 ):
-    await update_wallet(wallet.wallet.id, new_name)
-    return {
-        "id": wallet.wallet.id,
-        "name": wallet.wallet.name,
-        "balance": wallet.wallet.balance_msat,
-    }
+    new_wallet = await create_wallet(user_id=wallet.wallet.user, wallet_name=new_name)
+    logger.debug(f"Created wallet {new_wallet.id} of user {wallet.wallet.user}")
+    return new_wallet
```
Comment (callebtc) [inline lnbits/core/views/api.py:140]: Is this related to this PR? What does it do?

**Comment by arcbtc** [comment]
> > its a good idea, but it exceeds the scope of this pr.
> > this pr is meant to get rid of the ?usr inside url. we can build upon this PR further down the line.
> 
> * it is important to maintain backwards compatibility, some people will not want to switch to `email+pass` login right away
> * other users might want to ditch immediately `user_id` only auth and only use `email+pass`, they need to specify this somehow
> * I agree that `nostr`, `google`, `twitter` can come with a second PR

Agreed, should be transitional. I also hate the idea of being the only lightning wallet you have to log in to, have discussed with DNI and session token will last forever unless user uses log out

**Comment by arcbtc** [comment]
I am also strongly against ditching regular key authentication completely. LNbits makes it easy to build on, and that should continue. With nodes and the like using key authentication, its not something we need ditch.

We have spoken about being able to optionally turn off key authentication.

File: lnbits/core/models.py [hunk 2]
```diff
@@ -79,6 +80,17 @@ class WalletTypeInfo:
     wallet: Wallet
 
 
+class createUser(BaseModel):
```
Comment (motorina0) [inline lnbits/core/models.py:83]: - class names must start with upper case

File: lnbits/core/views/user_api.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+async def user(user=Depends(login_manager)) -> User:
+    return user
+
+
+@user_router.post(
+    "/api/v1/login", description="Login to the API via the username and password"
+)
+async def login_endpoint(
+    response: Response,
+    username: str = Body(),
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:23]: - should be called `email`. See line 35: `user = await get_account_by_email(username)`

File: lnbits/core/models.py [hunk 4]
```diff
@@ -88,6 +100,14 @@ class User(BaseModel):
     admin: bool = False
     super_user: bool = False
 
+    def login(self, password: str) -> bool:
+        pwd_bytes: bytes = password.encode("utf-8")
+        if not self.password:
+            raise Exception("User does not have a password.")
```
Comment (motorina0) [inline lnbits/core/models.py:106]: - information leak. It should also say `"Incorrect login."`. Not more.

File: lnbits/decorators.py [hunk 5]
```diff
@@ -220,8 +243,20 @@ async def require_invoice_key(
...
+        g().user = user
+        return user
+
+    if not usr:
+        raise HTTPException(
+            status_code=HTTPStatus.NOT_FOUND,
+            detail="Not logged in or provided ?usr argument.",
+        )
+
+    g().user = await get_user(usr)
```
Comment (motorina0) [inline lnbits/decorators.py:259]: - why is this `g().user` required, why not use a local variable?

Comment (motorina0) [reply lnbits/decorators.py:259]: - the code is fragile, is a new `async` call is introduced then race conditions will occur
- by running this code:
```py
    try:
        g_user_id = g().user.id if g().user else None
        print('### before [{id, g_id]',usr, g_user_id)
    except:
        pass

    g().user = await get_user(usr) # global?
    await asyncio.sleep(1)

    g_user_id = g().user.id if g().user else None
    print('### after [{id, g_id]', usr == g_user_id, usr, g_user_id)
```

 - I have managed to get into an inconsistent state
```log
### after [{id, g_id] False bf590154da614dbc943bb1fed4cf2661 f9270eec46404a38a1fda2e8af0df34a
```

File: lnbits/decorators.py [hunk 6]
```diff
@@ -241,8 +276,8 @@ async def check_user_exists(usr: UUID4) -> User:
     return g().user
 
 
-async def check_admin(usr: UUID4) -> User:
-    user = await check_user_exists(usr)
+async def check_admin(req: Request, usr: Optional[str] = None) -> User:
```
Comment (motorina0) [inline lnbits/decorators.py:279]: - why not use `UUID4` for `usr`?
```suggestion
async def check_admin(req: Request, usr: Optional[UUID4] = None) -> User:
```

File: lnbits/decorators.py [hunk 7]
```diff
@@ -252,8 +287,8 @@ async def check_admin(usr: UUID4) -> User:
     return user
 
 
-async def check_super_user(usr: UUID4) -> User:
-    user = await check_admin(usr)
+async def check_super_user(req: Request, usr: Optional[str] = None) -> User:
+    user = await check_admin(req, usr)
```
Comment (motorina0) [inline lnbits/decorators.py:291]: - why not use UUID4 for usr?
```suggestion
async def check_super_user(req: Request, usr: Optional[UUID4] = None) -> User:
```

File: lnbits/decorators.py [hunk 8]
```diff
@@ -252,8 +287,8 @@ async def check_admin(usr: UUID4) -> User:
     return user
 
 
-async def check_super_user(usr: UUID4) -> User:
-    user = await check_admin(usr)
+async def check_super_user(req: Request, usr: Optional[str] = None) -> User:
```
Comment (motorina0) [inline lnbits/decorators.py:290]: - why not use UUID4 for usr?
```suggestion
async def check_super_user(req: Request, usr: Optional[UUID4] = None) -> User:
```

File: lnbits/core/crud.py [hunk 9]
```diff
@@ -28,12 +29,45 @@
...
     Wallet,
     WebPushSubscription,
+    createUser,
 )
 
 # accounts
 # --------
 
 
+async def create_user(data: createUser) -> User:
```
Comment (motorina0) [inline lnbits/core/crud.py:39]:
```suggestion
async def create_user(data: CreateUser) -> User:
```

File: lnbits/core/crud.py [hunk 10]
```diff
@@ -28,12 +29,45 @@
     User,
     Wallet,
     WebPushSubscription,
+    createUser,
```
Comment (motorina0) [inline lnbits/core/crud.py:32]:
```suggestion
    CreateUser
```

File: lnbits/core/views/user_api.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
+from typing import Optional
+
+from fastapi import APIRouter, Body, Depends, Response, status
+from fastapi.responses import JSONResponse
+
+from ..crud import create_user, get_account_by_email, get_user
+from ..models import User, createUser
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:7]:
```suggestion
from ..models import User, CreateUser
```

File: lnbits/core/views/user_api.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    "/api/v1/login", description="Login to the API via the username and password"
+)
+async def login_endpoint(
+    response: Response,
+    username: str = Body(),
+    password: str = Body(),
+    usr: Optional[str] = Body(None),
+) -> JSONResponse:
+    if usr:
+        user = await get_user(usr)
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:28]: - a request with the body as listed below will always produce a valid access token (if `usr` exists)
- this means that `usr` has to remain forever secret

```json
{
    "username":"xxx",
    "password": "fake",
    "usr": "bf885e9c94084489a62c1e730dd3b389"
}
```

File: lnbits/core/views/user_api.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        return JSONResponse(
+            {
+                "detail": [
+                    {
+                        "loc": ["body", "password_repeat"],
+                        "msg": "passwords do not match",
+                    }
+                ]
+            },
+            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
```
Comment (motorina0) [inline lnbits/core/views/user_api.py:72]: - why not 401?
```suggestion
            status_code=status.HTTP_401_UNAUTHORIZED,
```

File: lnbits/core/crud.py [hunk 14]
```diff
@@ -28,12 +29,45 @@
...
+async def create_user(data: createUser) -> User:
+    if await get_account_by_email(data.email):
+        raise Exception("user exists")
+
+    pwd_bytes = data.password.encode("utf-8")
+    salt = gensalt()
+    password = hashpw(pwd_bytes, salt)
+
+    user_id = uuid4().hex
+    await db.execute(
```
Comment (motorina0) [inline lnbits/core/crud.py:48]: - should not allow duplicate emails

File: lnbits/core/static/js/index.js [hunk 15]
```diff
@@ -7,10 +7,61 @@ new Vue({
         show: false,
         data: {}
       },
-      walletName: ''
+      usr: '',
+      username: '',
+      password: '',
+      password_repeat: '',
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:13]: - cammelCase preferred

File: lnbits/core/static/js/index.js [hunk 16]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+      walletName: '',
+      signup: false
     }
   },
   methods: {
+    register: function () {
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:19]: - why not use `LNbits.api.request()`?

File: lnbits/core/static/js/index.js [hunk 17]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+        .then(response => {
+          this.$q.cookies.set('access-token', response.data.access_token)
+          this.$q.localStorage.set('lnbits.token', response.data.access_token)
+          this.$q.localStorage.set('lnbits.usr', response.data.usr)
+          window.location.href = '/wallet'
+        })
+        .catch(LNbits.utils.notifyApiError)
+    },
+    login: function () {
+      axios({
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:38]: - why not use `LNbits.api.request()`?

File: lnbits/core/static/js/index.js [hunk 18]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+        .then(response => {
+          this.$q.cookies.set('access-token', response.data.access_token)
+          this.$q.localStorage.set('lnbits.token', response.data.access_token)
+          // this.$q.localStorage.set('lnbits.usr', response.data.usr)
+          window.location.href = '/wallet'
+        })
+        .catch(LNbits.utils.notifyApiError)
+    },
+    login_usr: function () {
+      axios({
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:52]: - why not use `LNbits.api.request()`?

File: lnbits/core/static/js/index.js [hunk 19]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+      })
+        .then(response => {
+          this.$q.cookies.set('access-token', response.data.access_token)
+          this.$q.localStorage.set('lnbits.token', response.data.access_token)
+          // this.$q.localStorage.set('lnbits.usr', response.data.usr)
+          window.location.href = '/wallet'
+        })
+        .catch(LNbits.utils.notifyApiError)
+    },
+    login_usr: function () {
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:51]: - cammelCase prefered

```suggestion
    loginUser: function () {
```

File: lnbits/core/static/js/index.js [hunk 20]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+        url: '/api/v1/register',
+        data: {
+          email: this.username,
+          password: this.password,
+          password_repeat: this.password_repeat
+        }
+      })
+        .then(response => {
+          this.$q.cookies.set('access-token', response.data.access_token)
+          this.$q.localStorage.set('lnbits.token', response.data.access_token)
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:31]: - is there a `this.$q.localStorage.set('lnbits.token')` somewhere? Where is this used?

File: lnbits/core/static/js/index.js [hunk 21]
```diff
@@ -7,10 +7,61 @@ new Vue({
...
+          this.$q.localStorage.set('lnbits.usr', response.data.usr)
```
Comment (motorina0) [inline lnbits/core/static/js/index.js:32]: - where is this value read?

File: lnbits/core/views/api.py (hunk 1 above)
Comment (motorina0) [reply lnbits/core/views/api.py:140]: +1

File: lnbits/settings.py [hunk 22]
```diff
@@ -292,6 +292,9 @@ class EnvSettings(LNbitsSettings):
     lnbits_commit: str = Field(default="unknown")
     super_user: str = Field(default="")
     version: str = Field(default="0.0.0")
+    login_secret: str = Field(
```
Comment (motorina0) [inline lnbits/settings.py:295]: - possible security vulnerability if users forget to specify the environment variable
- can it default to a derivation `superuser_id` for example?
  - this way the secret is unique and the same after restarts and also can be recovered from the DB

File: lnbits/core/views/generic.py [hunk 23]
```diff
@@ -164,22 +163,52 @@ async def extensions_install(
...
+    user: User = Depends(check_user_exists),
+    wal: Optional[UUID4] = None,
 ):
-    user_id = usr.hex
-    user = await get_user(user_id)
+    user_id = user.id
 
-    if not user:
+    if (
+        len(settings.lnbits_allowed_users) > 0
```
Comment (motorina0) [inline lnbits/core/views/generic.py:172]: - this logic could be a method on `settings`

File: lnbits/decorators.py [hunk 24]
```diff
@@ -6,14 +6,37 @@
...
 from lnbits.core.crud import get_user, get_wallet_for_key
 from lnbits.core.models import User, WalletType, WalletTypeInfo
 from lnbits.db import Filter, Filters, TFilterModel
 from lnbits.requestvars import g
 from lnbits.settings import settings
 
+# from pydantic.types import UUID4
+
+
+def is_user(request: Request):
```
Comment (motorina0) [inline lnbits/decorators.py:19]: - from the name of the function I expect for the return type to be a `Bool`
- this function is in fact a getter

File: lnbits/decorators.py [hunk 25]
```diff
@@ -6,14 +6,37 @@
...
+    user = request.state.user
+    if user is None:
+        raise HTTPException(401)
+    return user
+
+
+def is_admin(request: Request):
```
Comment (motorina0) [inline lnbits/decorators.py:26]: - has naming convention of boolean function

File: lnbits/decorators.py [hunk 26]
```diff
@@ -6,14 +6,37 @@
...
+    user = is_user(request)
+    if not user.is_admin and not user.super_user:
+        raise HTTPException(401)
+
+    return user
+
+
+def is_super_user(request: Request):
```
Comment (motorina0) [inline lnbits/decorators.py:34]: - has naming convention of boolean function

File: lnbits/decorators.py [hunk 27]
```diff
@@ -220,8 +243,20 @@ async def require_invoice_key(
...
+    if req.state.user:
+        user = await get_user(req.state.user.id)
+        assert user, "Logged in user has to exist."
+        g().user = user
+        return user
+
+    if not usr:
+        raise HTTPException(
+            status_code=HTTPStatus.NOT_FOUND,
+            detail="Not logged in or provided ?usr argument.",
```
Comment (motorina0) [inline lnbits/decorators.py:256]: - argument -> query parameter

File: lnbits/static/js/base.js [hunk 28]
```diff
@@ -406,6 +431,15 @@ window.windowMixin = {
     }
   },
   created: function () {
+    // crop usr from url and login that user in the background
+    if (window.location.href.search('usr=') !== -1) {
```
Comment (motorina0) [inline lnbits/static/js/base.js:435]:
```js
const urlParams = new URLSearchParams(window.location.search);
const usrParam = urlParams.get('usr');
```

File: lnbits/decorators.py (hunk 5 above)
Comment (dni) [reply lnbits/decorators.py:259]: i think we should rethink the usage of `g()` altogether, i dont see much point in it

File: lnbits/core/views/user_api.py (hunk 13 above)
Comment (dni) [reply lnbits/core/views/user_api.py:72]: its the `/register` endpoint it is public

File: lnbits/static/js/base.js (hunk 28 above)
Comment (dni) [reply lnbits/static/js/base.js]: awesome

File: lnbits/decorators.py (hunk 24 above)
Comment (dni) [reply lnbits/decorators.py:19]: it is a decorator, maybe calling them require_user and so on is better?

File: lnbits/core/static/js/index.js (hunk 21 above)
Comment (dni) [reply lnbits/core/static/js/index.js:38]: `$q.cookies` are sent with the request for auth.
`$q.localstorage` as far is a know ia bonus

File: lnbits/core/views/user_api.py (hunk 3 above)
Comment (dni) [reply lnbits/core/views/user_api.py:23]: my issue with email is, i dont want to expect an email from new users to create an account.

File: lnbits/core/crud.py (hunk 14 above)
Comment (dni) [reply lnbits/core/crud.py:47]: its checked on line, 39 above, what do you mean?

File: lnbits/core/views/user_api.py (hunk 12 above)
Comment (dni) [reply lnbits/core/views/user_api.py:28]: yes. it is meant for backward compatibility

File: lnbits/core/views/user_api.py [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+                {"error": "not found"}, status_code=status.HTTP_404_NOT_FOUND
+            )
+    else:
+        try:
+            user = await get_account_by_email(username)
+            if not user:
+                return JSONResponse(
+                    {"error": "not found"}, status_code=status.HTTP_404_NOT_FOUND
+                )
+            _ = await load_user(user.id)  # type: ignore
```
[the remark this answers is not in the document]
Comment (dni) [reply lnbits/core/views/user_api.py:40]: `load_user` populates `req.state.user` for fastapi login

File: lnbits/core/views/user_api.py (hunk 12 above)
Comment (motorina0) [reply lnbits/core/views/user_api.py:28]: - some users will not want to upgrade to email+pass
- some user will want only email+pass (no user login)
- we do need a setting for the allowed login modes
- this PR needs to be fully backwards compatible
- suggestion:

```py
settings.allowed_auth_methods = ["basic", "email", "pubkey", "..."]
```

- the admin can choose one or more of the above

File: lnbits/settings.py (hunk 22 above)
Comment (dni) [reply lnbits/settings.py]: you will need to help me here, that is over my head. keep in mind superuser can change

**Comment by motorina0** [comment]
If new users are not allowed (see https://github.com/lnbits/lnbits/pull/2017/) then the `Sign Up` flow should not be allowed (UI and API checks)

File: lnbits/core/static/js/index.js (hunk 21 above)
Comment (motorina0) [reply lnbits/core/static/js/index.js:38]: I mean this one:
```js
          this.$q.localStorage.set('lnbits.usr', response.data.usr)
```

Comment (dni) [reply lnbits/core/static/js/index.js:38]: ye it is not needed, because its also available one `window.user.id`

**Comment by motorina0** [comment]
Will be deprecated in favor of https://github.com/lnbits/lnbits/pull/2092
